### PR TITLE
Migrating ca-select - initial check-in for review

### PR DESCRIPTION
**Pull request by kevmarchant**

### CODE REVIEW

File: components/ca-select.js [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+                                this.root.querySelectorAll('option')[0].disabled = true;
+                            }
+
+                            opt.selected = (item.value === this.value);
+                                
+                            if (!this.isTouch && this.multiple) {
+
+                                opt.onmousedown = e => {
+
+                                    helpers.cancelEvent(e || event);
```
Comment (ElementForge) [inline components/ca-select.js]: helpers.cancelEvent should be replaced by adding the cancel-event module as a dependency

File: components/ca-select.js [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                                    self.checkChange.call(self, this);
+                                }
+
+                                this.root.onclick = e => {
+
+                                    helpers.cancelEvent(e || event);
```
Comment (ElementForge) [inline components/ca-select.js]: helpers.cancelEvent should be replaced by adding the cancel-event module as a dependency

File: components/ca-select.js [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+            el.setAttribute("data-selected", !selected);
+        };
+
+        /**
+         * A seemingly pointless method
+         * @returns {void} void
+         */
+        onupdate(data) {
+
+            helpers.trace(data);
```
Comment (ElementForge) [inline components/ca-select.js]: helpers.trace should be replaced by adding the trace module as a dependency

File: components/ca-select.js [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+            if (this.multiple) {
+
+                let results = [];
+
+                if (this.root) {
+
+                    const options = (!isTouch) ? this.root.querySelectorAll('[data-selected="true"]') : Array.prototype.slice.call(this.root.options).where('selected', true);
+
+                    helpers.forEach(options, item => {
```
Comment (ElementForge) [inline components/ca-select.js]: helpers.forEach should be put into a new for-each module and added as a dependency

File: components/ca-select.js [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                // grab hold of all the option elements
+                const self = this;
+
+                setTimeout(() => {
+
+                    const options = self.root && self.root.querySelectorAll('option');
+
+                    if (options) {
+
+                        helpers.forEach(options, item => {
```
Comment (ElementForge) [inline components/ca-select.js]: helpers.forEach should be put into a new for-each module and added as a dependency

File: components/ca-select.js [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+                                // then set the data-selected attribute for this option
+                                item.setAttribute('data-selected', true);
+                            }
+                        });
+                    }
+                }, 0);
+            }
+        }
+
+        get multiple() {
```
Comment (ElementForge) [inline components/ca-select.js]: do we need JSdoc comments on all getters and setters?

File: components/ca-select.js [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+         */
+        get value() {
+
+            if (this.multiple) {
+
+                const results = [];
+
+                if (this.root) {
+
+                    const options = (!this.isTouch) ? this.root.querySelectorAll('[data-selected="true"]') : Array.prototype.slice.call(this.root.options).where('selected', true);
```
Comment (joeburton) [inline components/ca-select.js]: need to remove where method...

File: components/ca-select.js [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+            if (this.root) {
+                this.root.value = value;
+            }
+
+            // if this a multi-select element
+            if (this.multiple) {
+
+                // TODO: handle set with style = checkbox
+                // grab hold of all the option elements
+                const self = this;
```
Comment (joeburton) [inline components/ca-select.js]: need to remove
const self = this;

File: components/ca-select.js [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+
+                setTimeout(() => {
+
+                    const options = self.root && self.root.querySelectorAll('option');
```
Comment (joeburton) [inline components/ca-select.js]: should refer to this instead of self.

File: components/ca-select.js [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+         * @returns {void} void
+         */
+        loadSchema() {
+
+            const self = this;
+            const schemaUrl = this.src;
+
+            return componentSupport.request({ url: schemaUrl, dataType: 'json' }).then(data => {
+
+                self.schema = data;
```
Comment (joeburton) [inline components/ca-select.js]: self.schema = data;
change to 
this.schema = data;

File: components/ca-select.js [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+        * Called when element is initialised
+        * @returns {void} void
+        */
+        createdCallback() {
+
+            const self = this;
+            this.isTouch = ('ontouchend' in document.documentElement);
+            this.selections = {};
+
+            switch (self.style) {
```
Comment (joeburton) [inline components/ca-select.js]: self.style
becomes
this.style

File: components/ca-select.js [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+
+                if (this.style !== 'checkbox') {
+
+                    // add default (unselectable) option
+                    createElement(this.root, 'option', { value: '', 'data-index': 0, selected: true }, defaultLabel);
+                }
+
+                if (this.style === 'checkbox') {
+
+                    this.rootCheckbox = createElement(self.root, 'div', {});
```
Comment (joeburton) [inline components/ca-select.js]: self.root
becomes
this.root

File: components/ca-select.js [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+                            }
+
+                            opt.selected = (item.value === this.value);
+
+                            if (!this.isTouch && this.multiple) {
+
+                                opt.onmousedown = e => {
+
+                                    cancelEvent(e || event);
+                                    self.checkChange.call(self, this);
```
Comment (joeburton) [inline components/ca-select.js]: self.checkChange.call(self, this);
become
this.checkChange.call(this);
I think!?!

File: components/ca-select.js [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+     */
+    class Select extends HTMLElement {
+
+        /**
+        * Called when element is initialised
+        * @returns {void} void
+        */
+        createdCallback() {
+
+            const self = this;
```
Comment (kevmarchant) [inline components/ca-select.js]: I very much want to do this... but it counts as refactoring at the moment :-/

File: components/ca-select.js [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+            switch (self.style) {
+
+                case 'checkbox':
+                    this.root = this;
+                    break;
+
+                default:
+                    this.root = createElement(this, 'select', {});
+                    this.root.multiple = this.multiple;
+                    this.root.onchange = this.onChangeHandler.bind(self);
```
Comment (nealeu) [inline components/ca-select.js]: That doesn't sound right.  this.onChangeHandler() would invoke it.

I think it should be

    this.root.onchange = this.onChangeHandler.bind(this);

or

    this.root.onchange = () => this.onChangeHandler();

Comment (kevmarchant) [inline components/ca-select.js]: I prefer option B... I'll change it to that

**Pull request merged**